android/src/main/kotlin/com/authgear/flutter/AuthgearPlugin.kt

```diff
@@ -367,17 +373,63 @@ class AuthgearPlugin: FlutterPlugin, ActivityAware, MethodCallHandler, PluginReg
+    // When the app is re-installed, and restored from a backup.
+    // A new master key is created, but it cannot decrypt the restored EncryptedSharedPreferences.
+    // This problem is persistence until the EncryptedSharedPreferences is deleted.
+    //
+    // The official documentation of EncryptedSharedPreferences tell us to
+    // exclude the EncryptedSharedPreferences from a backup.
+    // But defining a backup rule is not very appropriate in a SDK.
+    // So we try to fix this in our code instead.
+    //
+    // This fix is tested against security-crypto@1.1.0-alpha06 and tink-android@1.8.0
+    // Upgrading to newer versions may result in the library throwing a different exception that we fail to catch,
+    // making this fix buggy.
+    //
+    // To reproduce the problem, you have to follow the steps here https://developer.android.com/identity/data/testingbackup#TestingBackup
+    // The example app has been configured to back up the EncryptedSharedPreferences and nothing else.
+    // One reason is to reproduce the problem, and another reason is that some platform, some Flutter,
+    // store large files in the data directory. That will prevent the backup from working.
+    //
+    // The fix is to observe what exception was thrown by the underlying library
+    // when the problem was re-produced.
+    // When we catch the exception, we delete the EncryptedSharedPreferences and re-create it.
+    //
+    // Some references on how other fixed the problem.
+    // https://github.com/stytchauth/stytch-android/blob/0.23.0/0.1.0/sdk/src/main/java/com/stytch/sdk/common/EncryptionManager.kt#L50
+    // https://github.com/tink-crypto/tink-java/issues/23
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
+      context.deleteSharedPreferences(name)
+    } else {
+      context.getSharedPreferences(name, Context.MODE_PRIVATE).edit().clear().apply()
+      val dir = File(context.applicationInfo.dataDir, "shared_prefs")
+      File(dir, "$name.xml").delete()
+    }
+  }
+
   private fun getSharePreferences(): SharedPreferences {
     val context = pluginBinding?.applicationContext!!
     if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
       val masterKey = MasterKey.Builder(context).setKeyScheme(MasterKey.KeyScheme.AES256_GCM).build()
-      return EncryptedSharedPreferences.create(
-        context,
-        "authgear_encrypted_shared_preferences",
-        masterKey,
-        EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
-        EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
-      )
+      try {
+        return EncryptedSharedPreferences.create(
+          context,
+          ENCRYPTED_SHARED_PREFERENCES_NAME,
+          masterKey,
+          EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
+          EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
+        )
+      } catch (e: InvalidProtocolBufferException) {
```

Comment:
I suggest we catch both `IOException` and `GeneralSecurityException` here, which are documented to be possible exceptions of `EncryptedSharedPreferences.create`.
https://developer.android.com/reference/androidx/security/crypto/EncryptedSharedPreferences#create(android.content.Context,java.lang.String,androidx.security.crypto.MasterKey,androidx.security.crypto.EncryptedSharedPreferences.PrefKeyEncryptionScheme,androidx.security.crypto.EncryptedSharedPreferences.PrefValueEncryptionScheme)

And it seems calling `deleteSharedPreferences` is always make sense when this method throwed execption.

Reply:
Fixed.